### ISSUE TITLE
Table of Contents with different orientation, font style and size 

### ISSUE THREAD

**Issue by CayhanBoran** [py-pdf/fpdf2#1122]
Hey,
I am trying to create a table of contents with different font style for landscape format. The aligments of dots in the texts are not aligning. I tried different formats, but the only best option is using the Courier format.

used function for table of contents:

`    def _render_toc(self, pdf: FPDF, outline: List):
        """Method to write the table of contents with links"""
        pdf.set_text_color(0, 0, 0)
        pdf.set_font("Arial", "B", size=16)
        pdf.cell(150, 1, "Table of Contents", align="C", center=True)

        pdf.ln(10)
        pdf.set_font("Arial", size=12)
        for section in outline:

            link = pdf.add_link(page=section.page_number, x=80, y=0)
            print(pdf.epw)
            if pdf.get_y() >= 170:
                pdf.page += 1
                pdf.set_y(35)
                pdf.set_x(10)
                pdf.cell(w=pdf.epw,
                         h=pdf.font_size,
                         text=f'{" " * section.level * 2} {section.name} {"." * (60 - section.level*2 - len(section.name))} {section.page_number}',
                         new_x="LMARGIN",
                         new_y="NEXT",
                         align="L",
                         # Cell method wants string for link, but the add link method returns an integer as a link
                         # Linking is working like this, we can ignore this type error
                         link=link)  # type: ignore
            else:
                pdf.cell(w=pdf.epw,
                         h=pdf.font_size,
                         text=f'{" " * section.level * 2} {section.name} {"." * (60 - section.level*2 - len(section.name))} {section.page_number}',
                         new_x="LMARGIN",
                         new_y="NEXT",
                         align="L",
                         link=link)  # type: ignore`

If I make it with Courier:

<img width="578" alt="courier_font" src="https://github.com/py-pdf/fpdf2/assets/54483713/3e4a4435-698c-4663-9ebd-d1be4b4e9a4c">

If I make it with Arial:

<img width="578" alt="arial_font" src="https://github.com/py-pdf/fpdf2/assets/54483713/f535fa63-854f-4f12-875f-1568254b0b4c">

If you could help me and solve this issue of font style in table of contents, I would be really happy.
Thanks!

**Comment by Lucas-C** [comment]
Hi @CayhanBoran!

Basically your issue is that `Courrier` is a monospace (or fixed-width) font: https://en.wikipedia.org/wiki/Monospaced_font
Whereas `Arial` is not monospace, leading to varying alignements in the resulting table of content.

You could try using a render function that righ-align page numbers, like this:
```python
def render_toc_with_right_aligned_page_numbers(pdf, outline):
    pdf.set_font("Helvetica", size=16)
    for section in outline:
        link = pdf.add_link(page=section.page_number)
        pdf.cell(text=f'{" " * section.level * 2} {section.name}', link=link, new_x="LEFT")
        pdf.cell(text=f"{section.page_number}", link=link, w=pdf.epw, align="R")
        pdf.ln()
```

**Comment by CayhanBoran** [comment]
Yes solves partially, I wanted to use Helvetica or Arial with dots, but if that is not the case. I guess I will not use dots in the table of contents.
Thanks.

**Comment by gmischler** [comment]
The underlying issue is that using dots to align the page number is a very ugly hack, which was cobbled together decades ago, and never fixed.

Ideally, someone would submit a PR creating a new and more functional TOC class, which contains adaptable TocEntry instances.
Desirable features might be:

* Text part of each TocEntry implemented as a [TextRegion](https://py-pdf.github.io/fpdf2/TextRegion.html), so it can use varying fonts for multilingual entries, possible spanning several lines.
* Ability to vary the total width of the TOC, in absolute terms or relative to the page width.
* Right aligned page numbers column with separately configurable font.
* Multicolumn TOCs.
  * Autowrapping.
  * With configuration to wrap between individual entries, or between higher level chapter headings.
* Configurable visual element to connect text and page number columns.
  * Automatically adapting to the available width between text and page number.
  *  This could be dots, dashes, a line, some graphic, etc, in a configurable color (or nothing).
  *  With multiline entries, configurable whether the connection (and page number) vertically aligns with the top or bottom of the entry.
* All the bells and whistles I didn't think about.

Any takers?

**Comment by Lucas-C** [comment]
Just a note: following @gmischler comment, I renamed this issue into `Feature request: new and more functional TOC class` and added the `up-for-grabs` label.

**Comment by Lucas-C** [comment]
For reference, a new ToC implementation was implemented by @andersonhc in version 2.8.2:
https://py-pdf.github.io/fpdf2/DocumentOutlineAndTableOfContents.html#reference-implementation
